packages/hooks/use-disclosure/src/index.ts

```diff
@@ -44,11 +44,14 @@ export function useDisclosure(props: UseDisclosureProps = {}) {
     onOpenPropCallbackRef?.();
   }, [isControlled, onOpenPropCallbackRef]);
 
-  const onOpenChange = useCallback(() => {
-    const action = isOpen ? onClose : onOpen;
+  const onOpenChange = useCallback(
+    (isOpen: boolean) => {
+      const action = isOpen ? onOpen : onClose;
```

Comment:
Since `isOpen` is managed within `useDisclosure` whether it is controlled or uncontrolled, is it necessary to pass it as an argument?

Reply:
Yes, in this case its needed. Because when the  `onClose` is called its modifiying the internal state as I said earlier ( `isOpen` becomes `false`), but at this point the `isOpen` which is in `useDisclosure`  is `false`.
Initially the call was like this  `isOpen ? onClose : onOpen`, thats why it was working  when `onClose` used to get called. 
But now that will cause infinite render due to `useEffect`, thats why I reversed it and and added and argument.

Reply:
Isn't that the expected behavior?

1. You press the close button → `state.close` (internal state becomes `false`).
2. `onOpenChange` is called (at this point, `useDisclosure` state is still `true`).
3. Since `isOpen` is `true`, `onClose` gets called.
4. The `useDisclosure` state becomes `false` (syncing the internal state with useDisclosure).

Accepting an argument would be a breaking change, so it might be better to leave it as is for backward compatibility.

Reply:
Could you clarify in what cases the `useEffect` bug occurs?

Reply:
> Isn't that the expected behavior?
> 
> 1. You press the close button → `state.close` (internal state becomes `false`).
> 2. `onOpenChange` is called (at this point, `useDisclosure` state is still `true`).
> 3. Since `isOpen` is `true`, `onClose` gets called.
> 4. The `useDisclosure` state becomes `false` (syncing the internal state with useDisclosure).
> 
> Accepting an argument would be a breaking change, so it might be better to leave it as is for backward compatibility.



> 1. You press the open button → external state becomes true. That will remount modal Component.
> 2. `useOverlayTriggerState` will re-run with isOpen `true`. After this, `useEffect` will be called and it will call the outer onOpenExchange(which was received as prop to modal), without passing argument((at this point, `useDisclosure` state is still `true`).
> 3. Since `isOpen` is `true`, `onClose` gets called(at this point, `useDisclosure` state is becomes `false`, but internal state is true).
> 4. Now, you press the close button → `state.close` (internal state becomes `false`).
> 5. `onOpenChange` is called (at this point, `useDisclosure` state is `false`), which means `onOpen` gets called but here `onClose` needs to get called.

Ya, i thought that it might be a breaking change, but we are also passing state while calling onOpenChange from `useOverlayTriggerState`

> https://github.com/nextui-org/nextui/blob/8a33eabb2583202fcc8fbc33e8f2ed23bb45f1a4/packages/components/modal/src/use-modal.ts#L124

Reply:
Oh, so the `useEffect` added in `useModal` is the issue. Hmm, with the current implementation, it might introduce new bugs. Could you consider exploring alternative implementation methods?

Reply:
What do you think? @jrgarciadev